### ISSUE TITLE
Making our jars executable, a.k.a. choosing a main class

### ISSUE THREAD

**Issue by samohtj** [rsanchez-wsu/fa15-ceg3120#37]
As of now, the jars that are made when we build our app can't actually be run. The reason for this is that the jars' manifests have no main class specified, so the JRE doesn't know where the main method is. Adding this to our build process would be relatively simple, but there's still the question of which class the main method should be in. So, we need to decide where the main entry point for the program is going to be.

Right now it seems that the app will be launched from the LoginPopup class in common. The only problem with this is that it needs to know which UI to launch from there, without importing from the modules. We could go another route and have a different main class for each module, each one inside that module's package, but then we'd have to set the main class for each module individually.

Once we've got a main class selected, adding it to our build would be as simple as changing line 109 of our common.xml to this:

```
<jar jarfile="${build.dir}/${jarname}.jar" basedir="${classes.dir}">
    <manifest>
        <attribute name="Main-Class" value="edu.wright.cs.fa15.concon.whatever.Classname"/>
    </manifest>
</jar>
```

Or, if we choose to have the main class inside the module rather than in common, we'd have to add this to each modules build.xml somehow.

Something to think about. Any ideas or feedback would be appreciated.

**Comment by pquackenbush** [comment]
Are you talking about making a directly executable jar? Because if you navigate to the the build directory of one of the modules and execute 

``` bash
java -cp contractor_client.jar edu.wright.cs.fa15.ceg3120.concon.common.LoginPopUp
```

You'll find that the main in LoginPopUp does actually execute.  The problem I'm seeing here however is that the resources aren't being copied into the jar causing a NullPointerException right off the bat.

**Comment by pquackenbush** [comment]
As for the second concern you raised:
"How do we know which UI to launch and how do we do it without importing from the module?"

This is where the polymorphism comes into play.  When the user logs in, a UserAccount object is created and shipped off, the database does its magic and determines what kind of account it is and send back all the profile information for that user, when the object comes back from the network stack, we should have a sub-class of UserAccount.  Each of these sub-classes have a launchGui() method that will call the method responsible for creating the UI for the team that is dealing with that user type.
To launch the UI from within common without an import, you'll need to use Reflection to change a compile time dependency into a runtime dependency.  the compiler doesn't care if the package and class you supply don't actually exist, but they better be there at runtime.  This is what I'm currently using in HomeownerAccount to launch the customer client:

``` java
    @Override
    public void launchGui() {
        //(new CustomerClient()).buildGui();
        try {
            Class<?> homeowner = 
                    Class.forName("edu.wright.cs.fa15.ceg3120.concon"
                            + ".client.customer.CustomerClient");
            Method meth = homeowner.getMethod("buildGui", this.getClass());
            meth.invoke(homeowner.newInstance(), this);
        } catch (RuntimeException e) { 
            throw e;
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
```

I am noticing that unless we find a way to merge the jars into a single all encompassing jar, this won't work...

**Comment by pquackenbush** [comment]
So took some time to make sure I wasn't feeding you false information

after commenting out any reference to the image resources and logger classes, I was able to launch the customer client from the contractor_client.jar.

``` bash
java -cp contractor_client.jar;customer_client.jar edu.wright.cs.fa15.ceg3120.concon.common.LoginPopUp
```

This assumes that they are in the same directory, but I'm almost positive that if you supplied the relative path after the semicolon, you would get the same result.

**Comment by samohtj** [comment]
Yes, I'm talking about creating a directly runnable jar. You can obviously run the program just fine from the command line, however it requires you to specify the class to run from. It shouldn't be up to the end user to know or care about that, so it should be configured beforehand.

**Comment by pquackenbush** [comment]
We could solve this with a simple batch and bash script that contains the launch command. I know it's not as elegant, but it's simple, most people would understand how to use it, and it doesn't make the common.xml more cryptic than it already is.
